File: MyLeetCode/python/Binary Search Tree Iterator.py
###
# Time Complexity: hasNext: O(1) next: Amortized(in average) O(1) visit each node once
# Space Complexity: O(h)
###

# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stk = []
        cur = root
        while cur:
            self.stk.append(cur)
            cur = cur.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return len(self.stk) != 0

    def next(self):
        """
        :rtype: int
        """
        top = self.stk.pop()
        cur = top.right
        while cur:
            self.stk.append(cur)
            cur = cur.left
        return top.val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())

File: MyLeetCode/Leetcode/Binary Search Tree Iterator.cpp
//Inorder iterative solution
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root) {
        TreeNode* curr = root;
        while (curr) {
            stk.push(curr);
            curr = curr->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !stk.empty();
    }

    /** @return the next smallest number */
    int next() {
        TreeNode* tmp = stk.top();
        stk.pop();
        if (tmp->right) {
            TreeNode* curr = tmp->right;
            while (curr) {
                stk.push(curr);
                curr = curr->left;
            }
        }
        return tmp->val;
    }
    stack<TreeNode*> stk;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */


//Sean's solution
//http://www.pixelstech.net/article/1373856647-Binary-tree-iterator-algorithm
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
public:
    BSTIterator(TreeNode *root) {
        TreeNode *curr = root;
        while (curr)
        {
            stk.push(curr);
            curr = curr->left;
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !stk.empty();
    }

    /** @return the next smallest number */
    int next() {
        if (stk.empty())
        {
            return INT_MIN;
        }
        else
        {
            TreeNode *top = stk.top();
            TreeNode *curr;
            stk.pop();

            if (top->right)
            {
                curr = top->right;
                while (curr)
                {
                    stk.push(curr);
                    curr = curr->left;
                }
            }
            return top->val;
        }
    }
private:
    stack<TreeNode*>stk;
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */